File: LoadTesting/populate_ref_id_update_uin.py
import psycopg2
from psycopg2.extras import execute_values
from typing import Dict, List
import time

CONFIG = {
    'SOURCE_DB': {  # MOSIP IDGen DB Session
        'host': 'localhost',
        'port': 5433,
        'user': 'postgres',
        'password': 'password',
        'database': 'mosip_kernel'
    },
    'TARGET_DB': { # Social Registry DB Session
        'host': 'localhost',
        'port': 5432,
        'user': 'socialregistryuser',
        'password': 'password',
        'database': 'socialregistrydb'
    }
}


# CREATE DB SESSION
def get_db_session(config: Dict) -> psycopg2.extensions.connection:
    try:
        return psycopg2.connect(
            host=config['host'],
            port=config['port'],
            user=config['user'],
            password=config['password'],
            database=config['database']
        )

    except Exception as e:
        print(f"Failed to connect to database: {e}")
        raise


# GET UNUSED UINs FROM kernel.uin
def get_unused_uins(session: psycopg2.extensions.connection, BATCH_SIZE: int) -> List[str]:
    with session.cursor() as cur:
        cur.execute("""
            SELECT uin 
            FROM kernel.uin
            WHERE uin_status = 'UNUSED'
            LIMIT %s
        """, (BATCH_SIZE,))

        return [row[0] for row in cur.fetchall()]


# GET res_partner IDs WHERE ref_id IS NULL
def get_res_partners_without_ref_id(session: psycopg2.extensions.connection, BATCH_SIZE: int) -> List[int]:
    with session.cursor() as cur:
        cur.execute("""
            SELECT id
            FROM res_partner
            WHERE ref_id IS NULL
            LIMIT %s
        """, (BATCH_SIZE,))
        return [row[0] for row in cur.fetchall()]


# UPDATE ref_id IN res_partner
def update_ref_ids(session: psycopg2.extensions.connection, res_partner_ids: List[int], uins: List[str]) -> bool:
    try:
        with session.cursor() as cur:
            # Build the list of tuples (id, uin)
            update_values = list(zip(res_partner_ids, uins))
            execute_values(cur, """
                UPDATE res_partner rp
                SET ref_id = data.uin
                FROM (VALUES %s) AS data(id, uin)
                WHERE rp.id = data.id
            """, update_values)
        session.commit()
        return True
    except Exception as e:
        session.rollback()
        print(f"Failed to update ref_ids: {e}")
        return False


# UPDATE uin_status = ISSUED IN kernel.uin
def update_uin_status(session: psycopg2.extensions.connection, uins: List[str]) -> bool:
    try:
        with session.cursor() as cur:
            cur.execute("""
                UPDATE kernel.uin 
                SET uin_status = 'ASSIGNED'
                WHERE uin = ANY(%s)
            """, (uins,))
        session.commit()

        return True

    except Exception as e:
        session.rollback()
        print(f"Failed to update UIN status: {e}")
        return False


def main():
    source_session = None
    target_session = None

    TOTAL_RECORDS_TO_PROCESS = 40000
    BATCH_SIZE = 40000

    try:
        source_session = get_db_session(CONFIG['SOURCE_DB']) # MOSIP IDGen DB Session
        target_session = get_db_session(CONFIG['TARGET_DB']) # Social Registry DB Session

        total_processed = 0
        start_time = time.time()

        while True:
            # Calculate current batch size
            remaining_records = TOTAL_RECORDS_TO_PROCESS - total_processed
            current_batch_size = min(BATCH_SIZE, remaining_records)

            # Fetch res_partner records with ref_id is null
            res_partner_ids = get_res_partners_without_ref_id(target_session, current_batch_size)

            # No more records to process
            if not res_partner_ids:
                print("No more res_partner records without ref_id")
                break

            # Get UINs from source database
            uins = get_unused_uins(source_session, len(res_partner_ids))

            # No more UINs to assign
            if not uins:
                print("No more UINs available")
                break

            actual_batch_size = min(len(res_partner_ids), len(uins))
            print(f"Processing batch of {actual_batch_size} records")

            # Truncate lists to the actual batch size
            res_partner_ids = res_partner_ids[:actual_batch_size]
            uins = uins[:actual_batch_size]

            # Update ref_ids in res_partner
            if update_ref_ids(target_session, res_partner_ids, uins):
                # Update uin_status in source database
                if update_uin_status(source_session, uins):
                    total_processed += actual_batch_size
                    print(f"Successfully processed {actual_batch_size} records")
                else:
                    print("Failed to update uin_status, rolling back batch")
                    continue
            else:
                print("Failed to update ref_ids, skipping batch")
                continue

            # Elapsed and rate statistics
            elapsed_time = time.time() - start_time
            rate = total_processed / elapsed_time if elapsed_time > 0 else 0
            print(
                f"Progress: {total_processed}/{TOTAL_RECORDS_TO_PROCESS} records processed ({rate:.2f} records/second)")

            # Break if completed processing required records
            if total_processed >= TOTAL_RECORDS_TO_PROCESS:
                print("Reached target number of records to process")
                break

        print(f"Processing completed. Total records processed: {total_processed}")

    except Exception as e:
        print(f"An error occurred: {e}")
        raise

    finally:
        if source_session:
            source_session.close()
        if target_session:
            target_session.close()


if __name__ == "__main__":
    main()

File: LoadTesting/populate_ref_id_update_uin.go
package main

import (
    "database/sql"
    "fmt"
    "log"
    "sync"
    "time"

    "github.com/lib/pq"
)

type DBConfig struct {
    Host     string
    Port     int
    User     string
    Password string
    Database string
}

var CONFIG = map[string]DBConfig{
    "SOURCE_DB": {
        Host:     "localhost",
        Port:     5433,
        User:     "postgres",
        Password: "password",
        Database: "mosip_kernel",
    },
    "TARGET_DB": {
        Host:     "localhost",
        Port:     5432,
        User:     "socialregistryuser",
        Password: "password",
        Database: "socialregistrydb",
    },
}

func getDBSession(config DBConfig) (*sql.DB, error) {
    connStr := fmt.Sprintf(
        "host=%s port=%d user=%s password=%s dbname=%s sslmode=disable",
        config.Host, config.Port, config.User, config.Password, config.Database,
    )
    db, err := sql.Open("postgres", connStr)
    if err != nil {
        return nil, fmt.Errorf("failed to connect to database: %w", err)
    }
    // Verify the connection is good
    if err := db.Ping(); err != nil {
        return nil, fmt.Errorf("failed to ping database: %w", err)
    }
    return db, nil
}

func getUnusedUINs(config DBConfig, batchSize int) ([]string, error) {
    db, err := getDBSession(config)
    if err != nil {
        return nil, err
    }
    defer db.Close()

    query := `SELECT uin FROM kernel.uin WHERE uin_status = 'UNUSED' LIMIT $1`
    rows, err := db.Query(query, batchSize)
    if err != nil {
        return nil, fmt.Errorf("failed to execute query: %w", err)
    }
    defer rows.Close()

    var uins []string
    for rows.Next() {
        var uin string
        if err := rows.Scan(&uin); err != nil {
            return nil, fmt.Errorf("failed to scan row: %w", err)
        }
        uins = append(uins, uin)
    }
    if err := rows.Err(); err != nil {
        return nil, fmt.Errorf("row iteration error: %w", err)
    }
    return uins, nil
}

func getResPartnersWithoutRefID(config DBConfig, batchSize int) ([]int, error) {
    db, err := getDBSession(config)
    if err != nil {
        return nil, err
    }
    defer db.Close()

    query := `SELECT id FROM res_partner WHERE ref_id IS NULL LIMIT $1`
    rows, err := db.Query(query, batchSize)
    if err != nil {
        return nil, fmt.Errorf("failed to execute query: %w", err)
    }
    defer rows.Close()

    var ids []int
    for rows.Next() {
        var id int
        if err := rows.Scan(&id); err != nil {
            return nil, fmt.Errorf("failed to scan row: %w", err)
        }
        ids = append(ids, id)
    }
    if err := rows.Err(); err != nil {
        return nil, fmt.Errorf("row iteration error: %w", err)
    }
    return ids, nil
}

func updateRefIDs(db *sql.DB, resPartnerIDs []int, uins []string) error {
    if len(resPartnerIDs) != len(uins) {
        return fmt.Errorf("length of resPartnerIDs and uins must be equal")
    }

    tx, err := db.Begin()
    if err != nil {
        return fmt.Errorf("failed to begin transaction: %w", err)
    }
    defer func() {
        if err != nil {
            tx.Rollback()
        }
    }()

    query := `
        WITH data(id, ref_id) AS (
            SELECT unnest($1::int[]), unnest($2::text[])
        )
        UPDATE res_partner rp
        SET ref_id = data.ref_id
        FROM data
        WHERE rp.id = data.id;
    `
    _, err = tx.Exec(query, pq.Array(resPartnerIDs), pq.Array(uins))
    if err != nil {
        tx.Rollback()
        return fmt.Errorf("failed to execute update query: %w", err)
    }

    if err := tx.Commit(); err != nil {
        return fmt.Errorf("failed to commit transaction: %w", err)
    }
    return nil
}

func updateUINStatus(db *sql.DB, uins []string) error {
    tx, err := db.Begin()
    if err != nil {
        return fmt.Errorf("failed to begin transaction: %w", err)
    }
    defer func() {
        if err != nil {
            tx.Rollback()
        }
    }()

    query := `
        UPDATE kernel.uin
        SET uin_status = 'ASSIGNED'
        WHERE uin = ANY($1)
    `
    _, err = tx.Exec(query, pq.Array(uins))
    if err != nil {
        tx.Rollback()
        return fmt.Errorf("failed to execute update query: %w", err)
    }

    if err := tx.Commit(); err != nil {
        return fmt.Errorf("failed to commit transaction: %w", err)
    }
    return nil
}

func processChunk(
    resPartnerIDsChunk []int,
    uinsChunk []string,
    sourceDBConfig, targetDBConfig DBConfig,
    resultChan chan<- int,
) {
    startTime := time.Now()
    threadName := fmt.Sprintf("Goroutine-%d", time.Now().UnixNano())
    fmt.Printf("Thread %s started at %s\n", threadName, startTime.Format("2006-01-02 15:04:05"))

    defer func() {
        endTime := time.Now()
        fmt.Printf("Thread %s ended at %s\n", threadName, endTime.Format("2006-01-02 15:04:05"))
    }()

    sourceDB, err := getDBSession(sourceDBConfig)
    if err != nil {
        log.Printf("Exception in processChunk (sourceDB): %v\n", err)
        resultChan <- 0
        return
    }
    defer sourceDB.Close()

    targetDB, err := getDBSession(targetDBConfig)
    if err != nil {
        log.Printf("Exception in processChunk (targetDB): %v\n", err)
        resultChan <- 0
        return
    }
    defer targetDB.Close()

    if err := updateRefIDs(targetDB, resPartnerIDsChunk, uinsChunk); err != nil {
        log.Printf("Failed to update ref_ids: %v\n", err)
        resultChan <- 0
        return
    }

    if err := updateUINStatus(sourceDB, uinsChunk); err != nil {
        log.Printf("Failed to update UIN status: %v\n", err)
        resultChan <- 0
        return
    }

    resultChan <- len(resPartnerIDsChunk)
}

func main() {
    TOTAL_RECORDS_TO_PROCESS := 40000
    BATCH_SIZE := 40000
    SUB_BATCH_SIZE := 10000

    totalProcessed := 0
    startTime := time.Now()

    for {
        remainingRecords := TOTAL_RECORDS_TO_PROCESS - totalProcessed
        currentBatchSize := BATCH_SIZE
        if remainingRecords < BATCH_SIZE {
            currentBatchSize = remainingRecords
        }

        resPartnerIDs, err := getResPartnersWithoutRefID(CONFIG["TARGET_DB"], currentBatchSize)
        if err != nil {
            log.Printf("Error fetching res_partner_ids: %v\n", err)
            break
        }
        if len(resPartnerIDs) == 0 {
            fmt.Println("No more res_partner records without ref_id")
            break
        }

        uins, err := getUnusedUINs(CONFIG["SOURCE_DB"], len(resPartnerIDs))
        if err != nil {
            log.Printf("Error fetching UINs: %v\n", err)
            break
        }
        if len(uins) == 0 {
            fmt.Println("No more UINs available")
            break
        }

        actualBatchSize := len(resPartnerIDs)
        if len(uins) < actualBatchSize {
            actualBatchSize = len(uins)
        }
        fmt.Printf("Processing batch of %d records\n", actualBatchSize)

        resPartnerIDs = resPartnerIDs[:actualBatchSize]
        uins = uins[:actualBatchSize]

        type chunk struct {
            resPartnerIDsChunk []int
            uinsChunk          []string
        }
        var chunks []chunk
        for i := 0; i < actualBatchSize; i += SUB_BATCH_SIZE {
            end := i + SUB_BATCH_SIZE
            if end > actualBatchSize {
                end = actualBatchSize
            }
            chunks = append(chunks, chunk{
                resPartnerIDsChunk: resPartnerIDs[i:end],
                uinsChunk:          uins[i:end],
            })
        }

        var wg sync.WaitGroup
        resultChan := make(chan int, len(chunks))
        for _, c := range chunks {
            wg.Add(1)
            go func(c chunk) {
                defer wg.Done()
                processChunk(c.resPartnerIDsChunk, c.uinsChunk, CONFIG["SOURCE_DB"], CONFIG["TARGET_DB"], resultChan)
            }(c)
        }

        wg.Wait()
        close(resultChan)

        for processed := range resultChan {
            totalProcessed += processed
            fmt.Printf("Processed %d records in thread\n", processed)
        }

        elapsedTime := time.Since(startTime).Seconds()
        rate := float64(totalProcessed) / elapsedTime
        fmt.Printf(
            "Progress: %d/%d records processed (%.2f records/second)\n",
            totalProcessed, TOTAL_RECORDS_TO_PROCESS, rate,
        )

        if totalProcessed >= TOTAL_RECORDS_TO_PROCESS {
            fmt.Println("Reached target number of records to process")
            break
        }
    }

    fmt.Printf("Processing completed. Total records processed: %d\n", totalProcessed)
}
